File: api/src/main/java/org/pmiops/workbench/auth/ServiceAccounts.java
package org.pmiops.workbench.auth;

import com.google.auth.oauth2.GoogleCredentials;
import java.io.IOException;
import java.util.List;
import org.springframework.stereotype.Component;

/**
 * Handles functionality related to loading service account credentials and generating derived /
 * impersonated credentials.
 */
@Component
public class ServiceAccounts {

  /**
   * Returns an appropriate set of backend service credentials with the given set of scopes.
   *
   * <p>This method uses AppIdentityService to return an instance of scoped AppEngineCredentials
   * when running in an App Engine environment.
   *
   * <p>Unfortunately, if we use GoogleCredentials.getApplicationDefault() from within an App Engine
   * environment, the returned credentials will be an instance of ComputeEngineCredentials, which
   * doesn't support scoped access tokens. Frustratingly, the call to .createScoped will silently
   * proceed by doing nothing -- meaning we only learn about the error once an attempt to use these
   * credentials fails in a downstream service due to bad scopes.
   *
   * <p>See https://github.com/googleapis/google-auth-library-java/issues/272 and
   * https://github.com/googleapis/google-auth-library-java/issues/172 for reference; this seems to
   * be a common pain point for users of the com.google.auth.oauth2 library.
   *
   * @param scopes
   * @return
   * @throws IOException
   */
  public static GoogleCredentials getScopedServiceCredentials(List<String> scopes)
      throws IOException {
    GoogleCredentials credentials = GoogleCredentials.getApplicationDefault().createScoped(scopes);
    credentials.refreshIfExpired();
    return credentials;
  }

  /**
   * Retrieves an access token with the specified set of scopes derived from Workbench service
   * credentials.
   */
  public static String getScopedServiceAccessToken(List<String> scopes) throws IOException {
    return getScopedServiceCredentials(scopes).getAccessToken().getTokenValue();
  }

  public static String getServiceAccountEmail(String serviceAccountName, String projectId) {
    return String.format("%s@%s.iam.gserviceaccount.com", serviceAccountName, projectId);
  }
}

File: api/src/main/java/org/pmiops/workbench/auth/ServiceAccounts.kt
package org.pmiops.workbench.auth

import com.google.api.client.googleapis.util.Utils.getDefaultJsonFactory

import com.google.api.client.googleapis.auth.oauth2.GoogleCredential
import com.google.api.client.http.HttpTransport
import com.google.appengine.api.appidentity.AppIdentityService
import com.google.appengine.api.appidentity.AppIdentityServiceFactory
import java.io.IOException
import org.pmiops.workbench.config.WorkbenchEnvironment
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

/**
 * Handles functionality related to loading service account credentials and generating derived /
 * impersonated credentials.
 */
@Component
class ServiceAccounts @Autowired
constructor(private val httpTransport: HttpTransport) {

    val credentialBuilder: GoogleCredential.Builder
        get() = GoogleCredential.Builder()

    /**
     * Retrieves an access token for the Workbench server service account. This should be used
     * carefully, as this account is generally more privileged than an end user researcher account.
     */
    @Throws(IOException::class)
    fun workbenchAccessToken(workbenchEnvironment: WorkbenchEnvironment, scopes: List<String>): String {
        // When running locally, we get application default credentials in a different way than
        // when running in Cloud.
        if (workbenchEnvironment.isDevelopment) {
            val credential = GoogleCredential.getApplicationDefault().createScoped(scopes)
            credential.refreshToken()
            return credential.accessToken
        }
        val appIdentity = AppIdentityServiceFactory.getAppIdentityService()
        val accessTokenResult = appIdentity.getAccessToken(scopes)
        return accessTokenResult.accessToken
    }

    /**
     * Converts a service account Google credential into credentials for impersonating an end user.
     * This method assumes that the given service account has been enabled for domain-wide delegation,
     * and the given set of scopes have been included in the GSuite admin panel.
     *
     *
     * See docs/domain-delegation.md for more details.
     *
     * @param serviceAccountCredential
     * @param userEmail Email address of the user to impersonate.
     * @param scopes The list of Google / OAuth API scopes to be authorized for.
     * @return
     * @throws IOException
     */
    @Throws(IOException::class)
    fun getImpersonatedCredential(
            serviceAccountCredential: GoogleCredential, userEmail: String, scopes: List<String>): GoogleCredential {
        // Build derived credentials for impersonating the target user.
        val impersonatedUserCredential = credentialBuilder
                .setJsonFactory(getDefaultJsonFactory())
                .setTransport(httpTransport)
                .setServiceAccountUser(userEmail)
                .setServiceAccountId(serviceAccountCredential.serviceAccountId)
                .setServiceAccountScopes(scopes)
                .setServiceAccountPrivateKey(serviceAccountCredential.serviceAccountPrivateKey)
                .setServiceAccountPrivateKeyId(serviceAccountCredential.serviceAccountPrivateKeyId)
                .setTokenServerEncodedUrl(serviceAccountCredential.tokenServerEncodedUrl)
                .build()

        // Initiate the OAuth flow to populate the access token.
        impersonatedUserCredential.refreshToken()
        return impersonatedUserCredential
    }
}
